packages/plugin-ext/src/plugin/node/debug/debug.ts

```diff
@@ -107,181 +130,150 @@ export class DebugExtImpl implements DebugExt {
-            throw new Error(`Debug adapter contribution '${contributionId}' not found.`);
-        }
-
-        return contribution.getSchemaAttributes();
+    async $getSchemaAttributes(debugType: string): Promise<IJSONSchema[]> {
+        const contribution = this.debuggersContributions.get(debugType);
+        return contribution && contribution.configurationAttributes || [];
     }
 
-    async $getConfigurationSnippets(contributionId: string): Promise<IJSONSchemaSnippet[]> {
-        const contribution = this.contributions.get(contributionId);
-        if (!contribution) {
-            throw new Error(`Debug adapter contribution '${contributionId}' not found.`);
-        }
-
-        return contribution.getConfigurationSnippets();
+    async $getConfigurationSnippets(debugType: string): Promise<IJSONSchemaSnippet[]> {
+        const contribution = this.debuggersContributions.get(debugType);
+        return contribution && contribution.configurationSnippets || [];
     }
 
-    async $provideDebugConfigurations(contributionId: string, folder: string | undefined): Promise<theia.DebugConfiguration[]> {
-        const contribution = this.contributions.get(contributionId);
-        if (!contribution) {
-            throw new Error(`Debug adapter contribution '${contributionId}' not found.`);
+    async $provideDebugConfigurations(debugType: string, workspaceFolderUri: string | undefined): Promise<theia.DebugConfiguration[]> {
+        let result: theia.DebugConfiguration[] = [];
+
+        const providers = this.configurationProviders.get(debugType);
+        if (providers) {
+            for (const provider of providers) {
+                if (provider.provideDebugConfigurations) {
+                    result = result.concat(await provider.provideDebugConfigurations(this.toWorkspaceFolder(workspaceFolderUri)) || []);
+                }
+            }
         }
 
-        return contribution.provideDebugConfigurations(undefined);
+        return result;
     }
 
-    async $resolveDebugConfigurations(
-        contributionId: string,
-        debugConfiguration: theia.DebugConfiguration,
-        folder: string | undefined): Promise<theia.DebugConfiguration | undefined> {
+    async $resolveDebugConfigurations(debugConfiguration: theia.DebugConfiguration, workspaceFolderUri: string | undefined): Promise<theia.DebugConfiguration | undefined> {
+        let resolved = debugConfiguration;
 
-        const contribution = this.contributions.get(contributionId);
-        if (!contribution) {
-            throw new Error(`Debug adapter contribution '${contributionId}' not found.`);
+        const providers = this.configurationProviders.get(debugConfiguration.type);
+        if (providers) {
+            for (const provider of providers) {
+                if (provider.resolveDebugConfiguration) {
+                    resolved = await provider.resolveDebugConfiguration(this.toWorkspaceFolder(workspaceFolderUri), debugConfiguration) || resolved;
```

Comment:
Should not `resolved` be passed instead of  `debugConfiguration` that next provider enhance a configuration from the previous provider?

Also, VS Code APIs are not very good at nullability, `resolveDebugConfiguration` can return `null` and `undefined` in practice meaning that resolution failed. In this case iterating does not make sense and we should exist without checking other providers.

From docs for `resolveDebugConfiguration`: `The resolved debug configuration or undefined or null.`